ADR/user-specified-tokens-build.md

```diff
@@ -0,0 +1,64 @@
+# Enable users to configure tokens/credentials for a Build/Test Pipeline
+
+Date: 2023-03-03
+
+## Status
+
+Provisional
+
+
+## Context
+
+As a user, I would like to configure sensitive information/tokens/credentials for the build/test pipeline to consume as part of a partner task.
+
+
+## Decision
+
+### How would the sensitive information be persisted and made available to the PipelineRun ?
+
+1. The UI would take in sensitive information as form input and upload it to SPI using the user's authentication token. See https://github.com/redhat-appstudio/service-provider-integration-operator/blob/main/docs/USER.md#uploading-access-token-to-spi-using-kubernetes-secret .
+
+2. Subsequently, the UI would create an `SPIAccessTokenBinding` CR with the appropriate label in the `.spec.secret.labels` to make the `Secrets` available to the build service or the integration service.
+
+
+### How does the build service and the integration service discover the `Secrets` ?
+
+While uploading a token, the user should be able to choose the name of the Pipeline `parameter` from a UI dropdown. The dropdown would be a pre-populated and hard-coded list of names associated with known partner integrations. 
+
+The user should also be allowed to provide a parameter name of choice. The same could be used in a customized PipelineRun definition in the user's source code's .tekton/ directory.
+
+| Pipeline Parameter    | Contents of the Secret | Name of the Secret        |
+| --------------------  | ---------------------- | ------------------------  |
+| codecov-github-token  | *input from the user*  | *autogenerated*           | 
+| codecov-token-secret  | *input from the user*  | *autogenerated*           |
+| redhat-ocm-token      | *input from the user*  | *autogenerated*           | 
+| *input from the user* | *input from the user*  | *autogenerated*           | 
+
+
+The build-service or the integration-service, reponsible for generating the PipelineRun needs to be able to map the name of the PipelineRun `parameter` with that of the `Secret` created by the UI, with contents provided by the user. To help the build-service do so, the `secrets` would be labelled with the name of the parameter for the build-service to look up, by the UI. 
```

Comment:
How can we prevent malicious tasks from getting access to secrets in a workspace that they shouldn't? 
* Create a pattern of isolation across tasks
  * If we can enforce that tasks need to have a unique identifier/prefix, then we can ensure that there is adequate enough isolation between task maintainers
  * We would still want to enable task maintainers to reuse secrets across tasks, so we don't want to force too fine of a granularity
  * Tasks should not try to request secrets outside their namespace
* Enforce pattern of isolation
  * This could just be enforced on the task submission side with "due diligence"
  * An additional guarantee would be to have additional checks on the build- or integration-service side preventing secret mounting

Reply:
With customization it's not possible to prevent - user can manually modify the pipelinerun definition.

Reply:
Oof. this sounds like a large problem. Thinking out loud...

This would be a problem _today_ -- not just if we start accepting partner tasks. Can the build or integration service (whomever is responsible for untangling the mappings) not perform validation during the mapping process? If tasks are not referencing secrets _directly_ then the services can just resolve and provide all secrets "scoped" to a task? That wouldn't necessarily help because the tasks can still be customized. Can we prevent the pipeline definitions from modifying certain tasks or detect if modifications have been made to tasks, preventing secrets from being set if the task is not in "approved form"?

Reply:
We do plan on having enforcement around the level of privilege a PipelineRun workload could have 
1) by limiting how powerful the `pipeline` sa is.
2) by ensuring the user is not able to use a different 'sa' available in the namespace.

> How can we prevent malicious tasks from getting access to secrets in a workspace that they shouldn't?

By educating users about not putting random Tasks off the internet in their Git .tekton/ directory. Therefore, it is in the hands of the users to not introduce a malicious Task in their workspace.

If they did manage to do so out of a lapse in good decision-making, only the Components in the specific namespace would be impacted.

Reply:
>  If tasks are not referencing secrets directly then the services can just resolve and provide all secrets "scoped" to a task? That wouldn't necessarily help because the tasks can still be customized.

+1

> Can we prevent the pipeline definitions from modifying certain tasks or detect if modifications have been made to tasks, preventing secrets from being set if the task is not in "approved form"?

We can mostly do things around : 
a) validate the image used to run Task container
b) validate the bundle being used to reference the Task
c) validating the yaml isn't full-proof but definitely possible using stackrox-like solutions.

However, ultimately, if the binary in the Task's image tries to do something malicious, it needs to be stopped by the boundaries of the pipeline `sa` privileges.